internal/metrics/remote_write.go

```diff
@@ -0,0 +1,394 @@
+	err := w.client.Store(ctx, data)
+	switch err.(type) {
+	case remote.RecoverableError:
+		return RemoteRecoverableError{err}
+	default:
+		return err
+	}
+}
+
+func (r *RemoteWrite) Update(config models.DeviceConfigurationMessage) error {
+	// the following two variables are only here until we implement reading form device configuration
+	serverURL := os.Getenv("REMOTEWRITE_SERVER_URL")
+	serverTimeout, _ := time.ParseDuration(os.Getenv("REMOTEWRITE_SERVER_TIMEOUT"))
+	if serverTimeout == 0 {
+		serverTimeout = time.Second * 10
+	}
+
+	// create configuration
+	r.config = RemoteWriteConfiguration{
+		Url:     serverURL,
+		Timeout: serverTimeout,
+	}
+
+	if r.config.Url == "" { // feature is disabled
+		return nil
+	}
+
+	// create client
+	if r.Client == nil {
+		serverURL, err := url.Parse(r.config.Url)
+		if err != nil {
+			return err
+		}
+		client, err := remote.NewWriteClient(r.deviceID, &remote.ClientConfig{
+			Timeout: model.Duration(r.config.Timeout),
+			URL: &config_util.URL{
+				URL: serverURL,
+			},
+		})
+		if err != nil {
+			return err
+		}
+		r.Client = &writeClient{
+			client: client,
+		}
+	}
+
+	return nil
+}
+
+func (r *RemoteWrite) Init(config models.DeviceConfigurationMessage) error {
+	err := r.Update(config)
+	go r.writeRoutine()
+	return err
+}
+
+// writeRoutine
+// Used as the goroutine function for handling ongoing writes to remote server
+func (r *RemoteWrite) writeRoutine() {
+	log.Infof("metric remote writer %s started", r.deviceID)
+	for {
+		r.Write()
+		time.Sleep(r.waitInterval)
+	}
+}
+
+// write
+// Writes all the metrics. Stops when either no more metrics or failure (after retries)
+func (r *RemoteWrite) Write() {
+
+	if r.Client == nil { // feature disabled
+		return
+	}
+
+	// start writing
+	for maxTSDB := r.tsdb.MaxTime(); maxTSDB.Sub(r.LastWrite) > 0; maxTSDB = r.tsdb.MaxTime() {
+
+		// set range start and end
+		minTSDB, err := r.tsdb.MinTime()
+		if err != nil {
+			log.Error("failed reading TSDB min", err)
+			return
+		}
+
+		rangeStart := r.LastWrite.Add(time.Millisecond) // TSDB time is in milliseconds
+		if rangeStart.Sub(minTSDB) < 0 {
+			rangeStart = minTSDB
+		}
+		rangeEnd := rangeStart.Add(r.RangeDuration)
+		if rangeEnd.Sub(maxTSDB) > 0 {
+			rangeEnd = maxTSDB
+		}
+
+		log.Infof("going to write metrics range %s(%d)-%s(%d). TSDB min max: %s(%d)-%s(%d)",
+			rangeStart.String(), rangeStart.UnixNano(), rangeEnd.String(), rangeEnd.UnixNano(),
+			minTSDB.String(), minTSDB.UnixNano(), maxTSDB.String(), maxTSDB.UnixNano(),
+		)
+
+		// read range
+		series, err := r.tsdb.GetMetricsForTimeRange(rangeStart, rangeEnd, true)
+		if err != nil {
+			if err != nil {
+				log.Errorf("failed reading metrics for range %d-%d", rangeStart.UnixNano(), rangeEnd.UnixNano())
+				return
+			}
+		}
+
+		// add device label
+		for _, s := range series {
+			s.Labels[DeviceLabel] = r.deviceID
+		}
+
+		// write range
+		if len(series) != 0 {
+			if r.writeRange(series) {
```

Comment:
Should we use a data limit here? We do not want to saturate network after the device was disconnected for extended period of time.

Reply:
writeRange breaks the range into RequestMaxSamples (just like ACM does).

Reply:
ok, can we make it configurable?

Reply:
sure. at the moment it is treated as an internal config and set by env var. I can add it to the DeviceConfig. Go for it?